### DISCUSSION TITLE
Easy Solution to wrap html-elements around custom parsed md/cm-elements?

### DISCUSSION THREAD

**Discussion by cyberpunkbln** [thephpleague/commonmark#635] [Q&A] [answered]
Hello, i use at the moment this lib with custom parsing commonmark into html with the CommonMarkConverter. Now i have a problem that i simple want the `<header>`-block around headlines(#). At the moment i write the `<header>` tag directly in commonmark.

How can i tell the lib that i want some html-blocks around the standard-parsing? Or should i make an new commonmark-tag?

Can anybody help me with a code-example? That will be nice.

thx

**Comment by colinodell** [comment, marked as answer]
You could create a [custom block renderer](https://commonmark.thephpleague.com/1.5/customization/block-rendering/) that replaces the [built-in heading renderer](https://github.com/thephpleague/commonmark/blob/1.5/src/Block/Renderer/HeadingRenderer.php).

For example, you could do something like this:

```php
<?php

use League\CommonMark\Block\Element\AbstractBlock;
use League\CommonMark\Block\Element\Heading;
use League\CommonMark\Block\Renderer\BlockRendererInterface;
use League\CommonMark\Block\Renderer\HeadingRenderer;
use League\CommonMark\ElementRendererInterface;
use League\CommonMark\HtmlElement;

final class MyCustomHeadingRenderer implements BlockRendererInterface
{
    private BlockRendererInterface $coreHeadingRenderer;

    public function __construct()
    {
        $this->coreHeadingRenderer = new HeadingRenderer();
    }

    /**
     * @param Heading                  $block
     * @param ElementRendererInterface $htmlRenderer
     * @param bool                     $inTightList
     *
     * @return string
     */
    public function render(AbstractBlock $block, ElementRendererInterface $htmlRenderer, bool $inTightList = false)
    {
        return '<heading>' . $this->coreHeadingRenderer->render($block, $htmlRenderer, $inTightLight) . '</heading'>;
    }
}
```

And then hook that up to the library like this:

```php
<?php

use League\CommonMark\Block\Element\Heading;
use League\CommonMark\Environment;

$environment = Environment::createCommonMarkEnvironment();

$myRenderer = new MyCustomHeadingRenderer();

$environment->addBlockRenderer(Heading::class, $myRenderer, 10);
```

I have not tested that code but I think it should work (or at least get you 90% of the way there)

**Comment by cyberpunkbln** [comment]
Thx for the code.

**Comment by cyberpunkbln** [comment]
The code functioned without error. I have a little new queston:

```
$environment = Environment::createCommonMarkEnvironment();
$myRenderer = new CustomRenderer();
$environment->addBlockRenderer(Heading::class, $myRenderer, 10);
$converter = new CommonMarkConverter();
```
What must i do that the CommonMarkConverter use the new Block-Rendering. At the moment nothing happended when is use this with Commonmarkconverter?

thx

**Comment by colinodell** [comment]
You need to pass the `$environment` into the constructor like this:

```php
$converter = new CommonMarkConverter([], $environment);
```

**Comment by cyberpunkbln** [comment]
it works, great thx
now i better understand code and examples

**Comment by cyberpunkbln** [comment]
Oh, its only the problem that he put "headings" about every <h*> blocks, give it chance that he only put "headings" around <h*>-blocks that are only in the start of the commonmark?

```
# i need heading
## i need heading, too

text

text

#### i need no more heading
```
